### ISSUE TITLE
Repeated database queries from BaseElement::getAreaRelationName()

### ISSUE THREAD

**Issue by kinglozzer** [silverstripe/silverstripe-elemental#928]
During element rendering, `forTemplate()` calls `getRenderTemplates()`, which in turn calls `getAreaRelationName()` to help build a list of applicable templates (presumably so that elements can be rendered differently based on which area they belong to).

`BaseElement::getAreaRelationName()` calls `$this->Parent()`, which runs through a typical flow for fetching a has_one component. However, these calls aren’t cached so this results in an additional database query for each element as it’s rendered:

https://github.com/silverstripe/silverstripe-elemental/blob/f626840fd6d1d4267e49014eaa55ae94bc490b85/src/Models/BaseElement.php#L758-L760

I’d suggest we replace this with something like:
```php
$class = DataObject::getSchema()->hasOneComponent($this, 'Parent');
$area = ($this->ParentID) ? DataObject::get_by_id($class, $this->ParentID) : null;
$area = $area ?: ElementalArea::singleton();
```

**Comment by michalkleiner** [comment]
Just a quick thought — wasn't there an enhancement from @mfendeksilverstripe around gettoppage or similar? Not sure if it applies here but it's the first thing that came to mind.

**Comment by mfendeksilverstripe** [comment]
You're right @michalkleiner , we do have a feature addressing this. It's opt in in `4` and built in on `master`. See [Top page feature](https://github.com/silverstripe/silverstripe-elemental/blob/4/src/TopPage/DataExtension.php#L19) for more details @kinglozzer.

**Comment by kinglozzer** [comment]
Unless I’m misunderstanding something, I’m not sure that’d help in this case. The code in question has already fetched the page, it’s another query _after_ that that then fetches an `ElementalArea` record that I’m trying to eliminate 😄

**Comment by mfendeksilverstripe** [comment]
You're right @kinglozzer, the Top page feature is focused on page lookup from blocks. I think it's reasonable to add a cached call for elemental area as well.

**Comment by GuySartorelli** [comment]
Pull request merged. It'll be automatically tagged by GitHub Actions.
